### PR TITLE
Add delegate transfer ownership handler

### PR DESCRIPTION
**Pull request by millken**
# Description
Transfering delegate ownership to a new address PART 2

Add an `identifier` to the candidate to represent its voting address. This field defaults to null. When a new action is processed, if the field is null, it will be set to the original owner. Similarly, when getting this value, if it is null, return its owner field

Other changes:
1. update `CandidateRegister` handler , the registered owner address cannot be the address in the `Identifier` list

Related PR:
https://github.com/iotexproject/iotex-core/pull/4236
https://github.com/iotexproject/iotex-core/pull/4239
https://github.com/iotexproject/iotex-core/pull/4245
https://github.com/iotexproject/iotex-core/pull/4250

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/candidate_transfer_ownership.go [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+	if !ok {
+		panic("fail to load the method")
+	}
+}
+
+// CandidateTransferOwnership is the action to transfer ownership of a candidate
+type CandidateTransferOwnership struct {
+	AbstractAction
+
+	name     string
```
Comment (dustinxie) [inline action/candidate_transfer_ownership.go:70]: don't need name

File: action/protocol/staking/stakingpb/staking.proto [hunk 2]
```diff
@@ -59,4 +59,9 @@ message BucketType {
 
 message Endorsement {
     uint64 expireHeight = 1;
+}
+
+message CandidateTransferOwnership {
+    float version = 1;
+    map<string, bytes> nameToOwner = 2;
```
Comment (millken) [inline action/protocol/staking/stakingpb/staking.proto:66]: map<string, string> = <new owner, old owner>?

File: action/protocol/staking/candidate_center.go [hunk 3]
```diff
@@ -28,6 +28,7 @@ type (
 		ownerMap         map[string]*Candidate
 		operatorMap      map[string]*Candidate
 		selfStkBucketMap map[uint64]*Candidate
+		ownerTransfers   map[string] /*name*/ address.Address /*newOwner*/
```
Comment (millken) [inline action/protocol/staking/candidate_center.go:31]: can be delete?

File: action/protocol/staking/candidate.go [hunk 4]
```diff
@@ -26,6 +26,7 @@ type (
 		Owner              address.Address
 		Operator           address.Address
 		Reward             address.Address
+		Voter              address.Address
```
Comment (CoderZhi) [inline action/protocol/staking/candidate.go:29]: what's voter? how does it work? what if the ownership be transferred for more than one time?

**Comment by CoderZhi** [comment]
write description please!

File: action/protocol/staking/candidate_center.go [hunk 5]
```diff
@@ -501,6 +524,16 @@ func (cb *candBase) getByName(name string) (*Candidate, bool) {
 	return d, ok
 }
 
+func (cb *candBase) getByIdentifier(identifier address.Address) (*Candidate, bool) {
+	candidates := cb.all()
+	for _, c := range candidates {
+		if c.Identifier != nil && address.Equal(c.Identifier, identifier) {
```
Comment (CoderZhi) [inline action/protocol/staking/candidate_center.go:530]: same here

File: action/protocol/staking/handler_candidate_transfer_ownership.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	}
+	//check if the new owner is self
+	if address.Equal(act.NewOwner(), caller) {
+		return &handleError{
+			err:           errors.New("new owner is the same as the current owner"),
+			failureStatus: iotextypes.ReceiptStatus_ErrUnauthorizedOperator,
+		}
+	}
+	//check the new owner is exist or is contract address
+	if acc, err := accountutil.LoadOrCreateAccount(csm.SM(), act.NewOwner()); err != nil || acc.IsContract() {
```
Comment (envestcc) [inline action/protocol/staking/handler_candidate_transfer_ownership.go:64]: I don't think it's necessary,  btw it also cannot reject future contract account

File: action/protocol/staking/handler_candidate_transfer_ownership.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+	_, fetchErr := fetchCaller(ctx, csm, big.NewInt(0))
+	if fetchErr != nil {
+		return log, nil, fetchErr
+	}
+
+	if err := p.validateCandidateTransferOwnership(ctx, act, csm, actCtx.Caller); err != nil {
+		return log, nil, err
+	}
+	candidate := csm.GetByOwner(actCtx.Caller).Clone()
+	if candidate.Identifier == nil || candidate.Identifier.String() == "" {
```
Comment (envestcc) [inline action/protocol/staking/handler_candidate_transfer_ownership.go:35]: I prefer the `Identifier` is already settled after get from csm

File: action/protocol/staking/candidate.go [hunk 8]
```diff
@@ -101,7 +104,7 @@ func (d *Candidate) isSelfStakeBucketSettled() bool {
 
 // Collision checks collsion of 2 candidates
 func (d *Candidate) Collision(c *Candidate) error {
-	if address.Equal(d.Owner, c.Owner) {
+	if address.Equal(d.GetIdentifier(), c.GetIdentifier()) {
```
Comment (envestcc) [inline action/protocol/staking/candidate.go:107]: Should still keep owner unique

File: action/protocol/staking/candidate_center.go [hunk 9]
```diff
@@ -520,22 +576,25 @@ func (cb *candBase) getBySelfStakingIndex(index uint64) (*Candidate, bool) {
...
 
-func (cb *candBase) collision(d *Candidate) (address.Address, address.Address, address.Address) {
+func (cb *candBase) collision(d *Candidate) (address.Address, address.Address, address.Address, address.Address) {
 	cb.lock.RLock()
 	defer cb.lock.RUnlock()
-	var name, oper, self address.Address
-	if c, hit := cb.nameMap[d.Name]; hit && !address.Equal(c.Owner, d.Owner) {
-		name = c.Owner
+	var name, owner, oper, self address.Address
+	if c, hit := cb.nameMap[d.Name]; hit && !address.Equal(c.GetIdentifier(), d.GetIdentifier()) {
```
Comment (envestcc) [inline action/protocol/staking/candidate_center.go:583]: keep collision on owner

File: action/protocol/staking/candidate.go [hunk 10]
```diff
@@ -178,16 +181,29 @@ func (d *Candidate) Deserialize(buf []byte) error {
 	return d.fromProto(pb)
 }
 
+// GetIdentifier returns the voter address
+func (d *Candidate) GetIdentifier() address.Address {
```
[the remark this answers is not in the document]
Comment (envestcc) [reply action/protocol/staking/candidate.go:189]: add a TODO, and will be fixed after the series of PRs merged, for reducing complexity of code rebase

File: action/protocol/staking/handler_candidate_transfer_ownership.go (hunk 6 above)
Comment (envestcc) [reply action/protocol/staking/handler_candidate_transfer_ownership.go:64]: after discuss offline, only transfer to an existed EOA

**Review by envestcc — APPROVED** [review]

File: action/protocol/staking/candidate.go [hunk 11]
```diff
@@ -212,6 +232,13 @@ func (d *Candidate) fromProto(pb *stakingpb.Candidate) error {
 		return err
 	}
 
+	if len(pb.GetIdentifierAddress()) > 0 {
```
Comment (dustinxie) [inline action/protocol/staking/candidate.go:235]:
```
if id := pb.GetIdentifierAddress(); len(id) > 0 {
   use id inside
}
```

File: action/protocol/staking/candidate_center.go [hunk 12]
```diff
@@ -292,16 +312,20 @@ func (m *CandidateCenter) collision(d *Candidate) error {
 		return err
 	}
 
-	name, oper, self := m.base.collision(d)
-	if name != nil && !m.change.containsOwner(name) {
+	name, owner, oper, self := m.base.collision(d)
+	// fmt.Printf("name=%s owner=%s oper=%s self=%s\n", name, owner, oper, self)
```
Comment (dustinxie) [inline action/protocol/staking/candidate_center.go:316]: remove this

File: action/protocol/staking/candidate_center.go [hunk 13]
```diff
@@ -195,8 +196,10 @@ func (m *CandidateCenter) ContainsOwner(owner address.Address) bool {
 		return true
 	}
 
-	_, hit := m.base.getByOwner(owner.String())
-	return hit
+	if d, hit := m.base.getByOwner(owner.String()); hit {
+		return !m.change.containsIdentifier(d.GetIdentifier())
+	}
+	return false
```
Comment (dustinxie) [inline action/protocol/staking/candidate_center.go:202]: do we need a `ContainsIdentifier()`?

File: action/protocol/staking/candidate_center.go [hunk 14]
```diff
@@ -390,7 +426,16 @@ func (cc *candChange) getByOwner(owner address.Address) *Candidate {
...
+	for _, d := range cc.dirty {
+		if address.Equal(owner, d.Owner) {
+			return d.Clone()
+		}
+	}
+	return nil
+}
+
+func (cc *candChange) getByIdentifier(identifier address.Address) *Candidate {
+	if d, ok := cc.dirty[identifier.String()]; ok {
```
Comment (dustinxie) [inline action/protocol/staking/candidate_center.go:438]: if identifier == nil {
		return nil
	}

**Review by dustinxie — APPROVED** [review]

**Comment by dustinxie** [comment]
pls fix before merge

**Pull request merged**